src/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs

```diff
@@ -221,11 +259,206 @@ public class ExpressionCreator<TDelegate>
     {
         public static LambdaExpression CreateExpressionFunc(Expression body, string name, bool tailCall, ReadOnlyCollection<ParameterExpression> parameters)
         {
-            return new Expression<TDelegate>(body, name, tailCall, parameters);
+            if (name == null && !tailCall)
+            {
+                switch (parameters.Count)
+                {
+                    case 0: return new Expression0<TDelegate>(body);
+                    case 1: return new Expression1<TDelegate>(body, parameters[0]);
+                    case 2: return new Expression2<TDelegate>(body, parameters[0], parameters[1]);
+                    case 3: return new Expression3<TDelegate>(body, parameters[0], parameters[1], parameters[2]);
+                    default: return new ExpressionN<TDelegate>(body, parameters);
+                }
+            }
+
+            return new FullExpression<TDelegate>(body, name, tailCall, parameters);
         }
     }
 #endif
 
+    internal sealed class Expression0<TDelegate> : Expression<TDelegate>
+    {
+        public Expression0(Expression body)
+            : base(body)
+        {
+        }
+
+        protected internal override int ParameterCount => 0;
+
+        protected internal override ParameterExpression GetParameter(int index)
+        {
+            throw new InvalidOperationException();
```

Comment:
`throw Error.ArgumentOutOfRange(nameof(index));` otherwise a out of range index to `Parameters` getting this far will have a different exception to before.